whipper/image/table.py

```diff
@@ -221,7 +220,7 @@ def getTrackEnd(self, number):
             # if on a session border, subtract the session leadin
             thisTrack = self.tracks[number - 1]
             nextTrack = self.tracks[number]
-            if nextTrack.session > thisTrack.session:
+            if int(nextTrack.session or 1) > int(thisTrack.session or 1):
```

Comment:
This looks like it fixes an issue unrelated to porting to Python 3… ?

Reply:
This fixes the last failing test:

```python
FAIL: testDuration (whipper.test.test_image_toc.CapitalMergeTestCase)
testDuration
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/twisted/internet/defer.py", line 151, in maybeDeferred
    result = f(*args, **kw)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/twisted/internet/utils.py", line 221, in runWithWarningsSuppressed
    reraise(exc_info[1], exc_info[2])
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/twisted/python/compat.py", line 464, in reraise
    raise exception.with_traceback(traceback)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/twisted/internet/utils.py", line 217, in runWithWarningsSuppressed
    result = f(*a, **kw)
  File "/home/travis/build/whipper-team/whipper/whipper/test/test_image_toc.py", line 271, in testDuration
    self.assertEqual(self.table.getFrameLength(), 173530)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/twisted/trial/_synctest.py", line 432, in assertEqual
    super(_Assertions, self).assertEqual(first, second, msg)
  File "/opt/python/3.5.6/lib/python3.5/unittest/case.py", line 829, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/opt/python/3.5.6/lib/python3.5/unittest/case.py", line 822, in _baseAssertEqual
    raise self.failureException(msg)
twisted.trial.unittest.FailTest: 184930 != 173530

----------------------------------------------------------------------
Ran 110 tests in 6.344s

FAILED (failures=1)
```
Extract from the first commit of this pull request (da9eb0797442aca7d38214a5af38d404b487b18d):
```diff
diff --git a/whipper/image/table.py b/whipper/image/table.py
index 57377a9..d196835 100644
--- a/whipper/image/table.py
+++ b/whipper/image/table.py
@@ -221,9 +220,10 @@ def getTrackEnd(self, number):
             # if on a session border, subtract the session leadin
             thisTrack = self.tracks[number - 1]
             nextTrack = self.tracks[number]
-            if nextTrack.session > thisTrack.session:
-                gap = self._getSessionGap(nextTrack.session)
-                end -= gap
+            if None not in [thisTrack.session, nextTrack.session]:
+                if nextTrack.session > thisTrack.session:
+                    gap = self._getSessionGap(nextTrack.session)
+                    end -= gap
 
         return end
```

It fails because if either `nextTrack.session` or `thisTrack.session` are `None` the `if` is false and the instructions inside it aren't executed. The check for `None` is needed because Python 3 doesn't allow `NoneType` comparisons (in Python 2 that was possible). IIRC correctly in that test `nextTrack.session` has value `2` while `thisTrack.session` is `None`. That means the Python 2 version evaluates the `if` condition to true, while the Python 3 version in the first commit does not.
This is the change I've introduced to fix the issue:
```diff
diff --git a/whipper/image/table.py b/whipper/image/table.py
index 22a6414..43553b2 100644
--- a/whipper/image/table.py
+++ b/whipper/image/table.py
@@ -220,10 +220,9 @@ def getTrackEnd(self, number):
             # if on a session border, subtract the session leadin
             thisTrack = self.tracks[number - 1]
             nextTrack = self.tracks[number]
-            if None not in [thisTrack.session, nextTrack.session]:
-                if nextTrack.session > thisTrack.session:
-                    gap = self._getSessionGap(nextTrack.session)
-                    end -= gap
+            if int(nextTrack.session or 1) > int(thisTrack.session or 1):
+                gap = self._getSessionGap(nextTrack.session)
+                end -= gap
 
         return end
```
Both the values of `nextTrack.session` and `thisTrack.session` are compared as int (if `None`, the value `1` is used for the comparison - as in disc session 1).